Add functions for figurate numbers & abundant, deficient, and perfect numbers

kellyi opened the pull request:
Hi!

I was working on writing a little library of number theory methods and came across this one, so I thought I'd just make a pull request to add a few of them.

This PR adds some functions to check for figurate numbers: triangle, square, pentagon, hexagon, heptagon, and octagon numbers. Each function returns a Boolean. I've also added functions to check for abundant, deficient, and perfect numbers -- also returning true/false.

Best,
Kelly

kellyi committed:
feat: add some functions for figurate numbers

add functions to check for:
- triangular numbers
- square numbers
- pentagonal numbers
- hexagonal numbers
- heptagonal numbers
- octagonal numbers

add functions to check for perfect, abundant, and deficient numbers
update index.js & README with new function names and documentation

kellyi committed:
fix: remove junk line

rsandor commented on lib/is_hexagonal.js:
```diff
@@ -0,0 +1,13 @@
+  */
+
+module.exports = function isHexagonal(n) {
+  return ((Math.sqrt((8 * n) + 1) + 1) / 4) % 1 === 0;
```
Perhaps a quick line note for the proof of this

rsandor commented on lib/is_octagonal.js:
```diff
@@ -0,0 +1,13 @@
+  */
+
+module.exports = function isOctagonal(n) {
+  return (((Math.sqrt((3 * n) + 1)) + 1) / 3) % 1 == 0;
```
Same here, add a comment to where you got the formula (proof would be great). Ditto for the rest of these. Mostly I'd like to make reading the source educational :)

rsandor commented on lib/is_perfect.js:
```diff
@@ -0,0 +1,20 @@
+  */
+
+module.exports = function isDeficient(n) {
+  var divisorsOfNumber = divisors(n);
```
Is there a faster algorithm known for this? I kinda doubt it but divisors is expensive :( worth checking out if you haven't already

rsandor commented:
@kellyi - great work! These were definitely needed! Left a few comments, please ping me back with your findings or if you need further help. After that we will get it merged and versioned.

Thank you so much!

rsandor commented:
@kellyi - Also, would you be so kind as to add unit tests to the `test/` directory for each of the new functions?

kellyi committed:
feat: add tests and document formulae

- add unit tests for each new method
- write explanations of how algos for isPentagonal, isOctagonal, etc
  work and include link to Wikipedia page with the formula for finding
  each figurate root
- fix alphabetical order in README.md

kellyi commented:
Hi @rsandor -- 

Sure! I wrote tests for each of these new methods and included them in the `test/` directory. I also added a bit of explanation for how those algorithms worked: essentially they implement formula I found on the Wikipedia articles for each kind of figurate number, and they work by finding whatever figurate root of the given number then checking whether or not the result's a whole number. I linked the relevant Wikipedia articles for each, and they typically look like this:

https://en.wikipedia.org/wiki/Octagonal_number

(I implemented something like the "Test for Octagonal Number" for each of the figurates.)

I'm not sure whether there's a faster algorithm for the abundant/deficient/perfect number's thing -- I'd love to see it if so -- BUT the RosettaCode site looks to have a possibly quicker implementation of finding the divisors of an Int. I'm not sure whether that code's licensed under MIT or GPL or what, so I didn't directly include it, but it might be worth checking out regardless:

http://rosettacode.org/wiki/Factors_of_an_integer#JavaScript

Glad these look to be helpful! Thanks for creating the library in the first place! Let me know if you'd like me to make any other changes or updates or anything.

Best,
Kelly

rsandor commented:
@kellyi - Great, I'll look at the changes when I am off work and hopefully get this merged tonight. Cheers :dancers:

rsandor commented:
@kellyi - Merged and released in [v1.1.0](https://github.com/rsandor/number-theory/releases/tag/v1.1.0). Cheers!

kellyi commented:
@rsandor Awesome. Thanks! :+1:
